nutils/topology.py

```diff
@@ -1479,10 +1479,9 @@ def __or__(self, other):
 
   @property
   def connectivity(self):
-    mask = numpy.array([bool(ref) for ref in self.refs] + [False]) # trailing false serves to map -1 to -1
-    renumber = numpy.cumsum(mask)-1
-    renumber[~mask] = -1
-    return tuple(types.frozenarray(renumber.take(ioppelems).tolist() + [-1] * (ref.nedges - len(ioppelems))) for ref, ioppelems in zip(self.refs, self.basetopo.connectivity) if ref)
+    renumber = numeric.invmap([i for i, ref in enumerate(self.refs) if ref], length=len(self.refs)+1, nan=-1) # length=..+1 serves to map -1 to -1
```

Comment:
I'm not sure if it is faster or more readable, but you can replace `[i for i, ref in enumerate(self.refs) if ref]` with `numpy.formiter(self.refs, dtype=bool).nonzero()[0]`.

Reply:
I prefer the current for readability, and would expect it to be faster as well since the same work is involved in forming the boolean.